src/test/java/com/eighthlight/fabrial/test/server/TcpServerEchoIntegrationTest.java

```diff
@@ -0,0 +1,47 @@
+package com.eighthlight.fabrial.test.server;
+
+import com.eighthlight.fabrial.test.client.EchoRequest;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+public class TcpServerEchoIntegrationTest extends TcpServerIntegrationTest {
+  @BeforeEach
+  void startAndConnect() throws IOException {
+    server.start();
+    client.connect();
+  }
```

Comment:
You've made yourself that nice seam for `EchoRequest` where it only depends on `InputStream` and `OutputStream`, but you're still testing up at this really high level. Isolated unit tests will be faster, more reliable, more targeted in their failure cases (i.e. a test failure tells you exactly where the production code bug lives). Check out `java.io.ByteArrayInputStream` and `java.io.ByteArrayInputStream` which are very useful for stream-based unit tests.

And it'd be good for future features to get into that unit testing mindset.

Reply:
Ahh, I see, you're not actually even *using* the server that you spun up! You're just using it to get ahold of 2 streams, and calling methods directly on the `EchoRequest`! It's *like* a unit test, but also adds the painful parts of integration testing! So you can simplify this a lot by just moving to BAIS/BAOS and losing the server startup pieces. This is showing a pain point of test inheritance - very easy to bring in more behavior than you need/want! 

I finally noticed this when I was looking for where `EchoRequest` was wired up to the  production code, and found that it wasn't! This was surprising since there was an integration test, but now that I see how the test is working more clearly, it makes sense that the full wiring isn't in place.

Reply:
> Ahh, I see, you're not actually even using the server that you spun up! You're just using it to get ahold of 2 streams, and calling methods directly on the EchoRequest

The server is being used. The I/O streams are coming from the client, which is connected to the server. If the server wasn't being used, there wouldn't be an echo response.

Reply:
> java.io.ByteArrayInputStream and java.io.ByteArrayInputStream which are very useful for stream-based unit tests.

This occurred to me too, could be very useful.

Reply:
Ahhhh, whoops, OK, sorry. I took another look, and I had previously interpreted `EchoRequest` as being the thing that did the job of echoing. But that's really done by the `is.transferTo(os);` line in `TcpServer`. My fault. So I'm back on board with your 2nd checkbox (refactoring the echo into a separate connection handler) since your design will need that. I'd think something that looks pretty similar to `EchoRequest`, without the serialization bits. And it seems like generally this current `EchoRequest` object might not have anything to do with echoing, and is more about just reading & writing some underlying streams? If that's true, could do with a rename.

Reply:
> But that's really done by the is.transferTo(os); line in TcpServer

✅ 

>  I'd think something that looks pretty similar to EchoRequest, without the serialization bits

Yeah, except the more I think about it, this kind of echo functionality isn't really compatible with HTTP. An HTTP echo would just respond with the same payload, whereas this server echoes literally everything. This puts me at an interesting cross-roads. Some of this code is likely to be obsolete once I start on HTTP in earnest, so I can try to focus on the connection-accepting aspects and error handling of `TcpServer` and essentially use the rest as temporary scaffolding.

Having said all that, I'm still going to try to come at this more pragmatically. Really hoping to get a good chunk of HTTP done this week.

Reply:
Yeah don’t burn a lot of time on the echo behavior, agreed - that assignment was more about proving sockets are hooked up OK.